File: dart/leet_code/623.dart
/// Dart implementation of the LeetCode problem 623: Add One Row to Tree.
///
/// This module contains a class `Solution` with a method `addOneRow`.
/// The `addOneRow` method takes in three parameters:
/// - `root`: A TreeNode object representing the root of the tree.
/// - `val`: An integer representing the value to be added to the new row.
/// - `depth`: An integer representing the depth at which the new row should be added.
///
/// The method returns a TreeNode object representing the root of the tree after adding the new row.
///
/// Error Handling:
/// - This method assumes that the input parameters are well-formed, i.e., `root` is a TreeNode object, `val` is an integer, and `depth` is a positive integer.
/// - If the input parameters are not well-formed, the behavior of the method is undefined.

class TreeNode {
  int val;
  TreeNode? left;
  TreeNode? right;

  /// Constructor for the TreeNode class.
  ///
  /// @param val The value of the node.
  /// @param left The left child of the node.
  /// @param right The right child of the node.
  TreeNode([this.val = 0, this.left, this.right]);
}

class Solution {
  /// Adds a new row to the tree at the specified depth.
  ///
  /// The method uses a stack to perform a depth-first search (DFS) on the tree from the root node.
  /// If the specified depth is reached, a new row is added to the tree.
  ///
  /// @param root The root of the tree.
  /// @param val The value to be added to the new row.
  /// @param depth The depth at which the new row should be added.
  /// @return A TreeNode object representing the root of the tree after adding the new row.
  TreeNode? addOneRow(TreeNode? root, int val, int depth) {
    // If the specified depth is 1, add a new root to the tree.
    if (depth == 1) return TreeNode(val, root, null);

    int cur = 1;
    List<TreeNode> stack = [root!];

    // Perform DFS on the tree until the specified depth is reached.
    while (stack.isNotEmpty && cur < depth - 1) {
      cur++;
      final len = stack.length;
      for (int i = 0; i < len; i++) {
        var temp = stack.removeAt(0);

        // If the current node has a left child, add it to the stack.
        if (temp.left != null) stack.add(temp.left!);

        // If the current node has a right child, add it to the stack.
        if (temp.right != null) stack.add(temp.right!);
      }
    }

    // Add a new row to the tree at the specified depth.
    final len = stack.length;
    for (int i = 0; i < len; i++) {
      var temp = stack.removeAt(0);
      var l = temp.left;
      var r = temp.right;
      temp.left = TreeNode(val, l, null);
      temp.right = TreeNode(val, null, r);
    }

    // Return the root of the tree after adding the new row.
    return root;
  }
}

File: rust/src/leet_code/623.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }

// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::cell::RefCell;
use std::rc::Rc;

// Defining a type alias for `Option<Rc<RefCell<TreeNode>>>` for convenience
type OptNode = Option<Rc<RefCell<TreeNode>>>;

/// This module contains the implementation of the `Solution` struct.
///
/// The `Solution` struct provides a method `add_one_row` which adds a row of nodes with a given value to a binary tree at a given depth.
/// The function takes an optional reference-counted, mutable reference to the root node of the tree, an integer value for the new nodes, and an integer depth.
/// It returns an optional reference-counted, mutable reference to the root node of the modified tree.
///
/// # Parameters
///
/// * `root` - An optional reference-counted, mutable reference to the root node of the tree.
/// * `val` - An integer value for the new nodes.
/// * `depth` - An integer depth at which to add the new row of nodes.
///
/// # Returns
///
/// * `OptNode` - An optional reference-counted, mutable reference to the root node of the modified tree.
///
/// # Errors
///
/// This function will return `None` if the root node is `None`.
impl Solution {
    pub fn add_one_row(root: OptNode, val: i32, depth: i32) -> OptNode {
        // If the root node is None, return None
        if root.is_none() {
            return None;
        }

        // Define a closure to create a new node with a given value and optional left and right child nodes
        let new_node = |left: OptNode, right: OptNode, val: i32| -> OptNode {
            Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
        };

        // If the root node is Some, borrow it mutably
        if let Some(ref rc_node) = root {
            let mut node = rc_node.borrow_mut();
            // Match on the depth
            match depth {
                // If the depth is 1, return a new node with the root node as its left child and None as its right child
                1 => {
                    return new_node(root.clone(), None, val);
                }
                // If the depth is 2, replace the left and right child nodes of the root node with new nodes
                2 => {
                    node.left = new_node(node.left.take(), None, val);
                    node.right = new_node(None, node.right.take(), val);
                }
                // If the depth is greater than 2, recursively call `add_one_row` on the left and right child nodes of the root node
                _ => {
                    Self::add_one_row(node.left.clone(), val, depth - 1);
                    Self::add_one_row(node.right.clone(), val, depth - 1);
                }
            }
        }
        // Return the root node
        root
    }
}